### DISCUSSION TITLE
Support (be compatible with) Fable

### DISCUSSION THREAD

**Discussion by sergey-tihon** [fsprojects/SwaggerProvider#28] [Ideas]
[Should be possible to use Swagger Provider from Fable](https://twitter.com/alfonsogcnunez/status/739385383465717760).

There are some issues with reading AST of provided types 
https://github.com/fsharp/FSharp.Compiler.Service/issues/568#issuecomment-223670266

P.S.: [Working RegExp provider](https://github.com/fsprojects/Fable/blob/master/src/providers/regex/src/Fable.Providers.Regex.fsx#L15-L16)

**Comment by sergey-tihon** [comment]
@alfonsogarciacaro where is the latest version on Fable TP that `works`?

**Comment by alfonsogarciacaro** [comment]
Hi @sergey-tihon! Fable 0.7 should work with Type Providers, the main problem with Fable 1.0 is it's a netcore app and TPs won't work with netcore until netstandard 2.0 is released (July?).

I've been thinking about giving support to the Swagger TP. At the beginning I thought it would be better to create a specific version of the TP for Fable, but now I'm considering it it's easier to just provide replacements for the classes needed by the Swagger TP (basically the `HttpClient` class I believe).

@sergey-tihon, what do you think? Should we start working of this even if it only works with Fable 0.7? Or may it wait until netstandard2.0 is released and TPs work with netcore?

**Comment by sergey-tihon** [comment]
I think that we should go with Fable 1.0 ... As I understand everything is unclear with type provider. netstandard 2.0 may help with erased type providers, but probably [not with generative](https://github.com/Microsoft/visualfsharp/issues/2406).

I've created template sample generative type provider https://github.com/sergey-tihon/generative-typeprovider-vnext with goal to find out the way of mastering type providers that work "everywhere".

I agree with you that TP should be designed with idea that emitted code may be executed in JS world and provider replacements on rough edges.

So I guess that we should wait for `netstandard 2.0` ... because in `Fable 1.0` world everything (here I mean TP design-time component) should be runnable on `.NET Core`. So I have time to learn Fable 😄

**Comment by stweb** [comment]
I stumbled upon this today. My workaround is using NSwag to create a TypeScript client from swagger.json, then using ts2fable...

**Comment by stweb** [comment]
Now that 2.0 is in preview 2, did anyone try making this work with .net core?

**Comment by alfonsogarciacaro** [comment]
There are still several problems with preview 2, specially with the SDK, so I would wait until preview 3 at least before trying anything. Also, the main problem to adapt the Swagger provider is it uses System.Http classes in the erasure code, which are not compatible with Fable at the moment. There are two possible solutions for this:

- Use conditional compilation (`#if FABLE_COMPILER`) to replace erasure code in the Swagger provider using Fetch API instead of System.Http ([example](https://github.com/fsprojects/FSharp.Data.GraphQL/blob/e19b4002901175c098b95eeb09f775993675bedd/src/FSharp.Data.GraphQL.Client/QuotationHelpers.fs#L51)).
- Add replacements to Fable for System.Http classes. This can be helpful for other user cases, but it'll probably be more work.

Unfortunately I don't have time after FableConf, but if someone wants to undertake this I can provide guidance.

**Reply by antoniolauro1** [reply]
Hello, what's the status of this issue? In my team we are interested in a Fable Swagger provider.  If not already solved,  we could possibly we willing to undertake this.

**Reply by sergey-tihon** [reply]
@antoniolauro1 It should be already possible with [project  Hawaii](https://github.com/Zaid-Ajaj/Hawaii)

**Reply by antoniolauro1** [reply]
Ok, thanks! I'll have a look at it.
